### PR TITLE
[4.0] Separate core.js into multiple files

### PR DESCRIPTION
**Pull request by HLeithner**
Do NO SQUASH merge this commit please.

There are 2 reasons for this PR. First get core.js more module and split into multiple files for better maintainability. The second is a copyright issue reported by @brianteeman for the webcomponents part of the core.js

### Summary of Changes
Make it possible to split concat javascript files into a main file based on the directory name.
Add change history in git commits to preserve copyright and change information.

### Testing Instructions
you need to run npm install,
test Joomla core functionality like
* javascript translations
* check error  messages creation in forms / ajax forms
* check webcomponents loaded correctly  

### Expected result
works

### Actual result
works

### CODE REVIEW

**Comment by brianteeman** [comment]
Thanks for dealing with this.

File: build/media_source/system/js/core.es6/message.es6 [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            messageContainer = document.getElementById('system-message-container');
+        } else {
+            messageContainer = document.querySelector(selector);
+        }
+
+        if (typeof keepOld === 'undefined' || (keepOld && keepOld === false)) {
+            Joomla.removeMessages(messageContainer);
+        }
+
+        [].slice.call(Object.keys(messages)).forEach((type) => {
```
Comment (C-Lodder) [inline build/media_source/system/js/core.es6/message.es6]: Probably a copy and paste but `[].slice.call` isnt needed here as Object.keys() already returns an array

Comment (HLeithner) [reply build/media_source/system/js/core.es6/message.es6]: It's copy paste and I would like to keep this PR only for this split and copyright issue. We can create a new PR afterwards.

File: build/media_source/system/js/core.es6/message.es6 [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+     *
+     * @param   {object}  xhr         XHR object.
+     * @param   {string}  textStatus  Type of error that occurred.
+     * @param   {string}  error       Textual portion of the HTTP status.
+     *
+     * @return  {object}  JavaScript object containing the system error message.
+     *
+     * @since  3.6.0
+     */
+  Joomla.ajaxErrorsMessages = (xhr, textStatus) => {
```
Comment (dgrammatiko) [inline build/media_source/system/js/core.es6/message.es6:178]: @HLeithner this doesn't belong here but rather in the Ajax.es6

File: build/media_source/system/js/core.es6/customevent.es6 [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/**
+ * @copyright  Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (dgrammatiko) [inline build/media_source/system/js/core.es6/customevent.es6:2]: The copyright here is wrong. The code belongs to Mozilla as it is noted in line 9

**Comment by dgrammatiko** [comment]
I think you are also missing this:
```js
       // Only define the Joomla namespace if not defined.
	window.Joomla = window.Joomla || {};
```

File: build/media_source/system/js/core.es6/customevent.es6 (hunk 3 above)
Comment (brianteeman) [reply build/media_source/system/js/core.es6/customevent.es6:2]: Yikes - if that is how you have been commenting that code is copyright someone else then @HLeithner and myself have to conduct an even bigger review to make sure we are not abusing someones copyright.

Luckily in this specific case the copyright statement is correct. 

All code samples on mdn are public domain 

> Code samples added on or after August 20, 2010 are in the public domain (CC0). No licensing notice is necessary, 

And it is changed significantly to add our own copyright

Comment (dgrammatiko) [reply build/media_source/system/js/core.es6/customevent.es6:2]: > And it is changed significantly to add our own copyright

No, the code was copy/paste. It's one of the countless polyfills than MDN has on their site (provided by **ALL** the browser vendors (thus the public domain)

Comment (brianteeman) [reply build/media_source/system/js/core.es6/customevent.es6:2]: I assume you are referring to this exact polyfill 
https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent#Polyfill

Maybe you copy & pasted it but it is significantly changed

File: build/media_source/system/js/core.es6/message.es6 (hunk 2 above)
Comment (HLeithner) [reply build/media_source/system/js/core.es6/message.es6:178]: I moved it thx

File: build/media_source/system/js/core.es6/customevent.es6 (hunk 3 above)
Comment (HLeithner) [reply build/media_source/system/js/core.es6/customevent.es6:2]: Brian has more knowledge about this copyright thing then me so I trust him and keep it the way it is. thx

**Comment by HLeithner** [comment]
> I think you are also missing this:
> 
> ```js
>        // Only define the Joomla namespace if not defined.
> 	window.Joomla = window.Joomla || {};
> ```

This is in the core.es6.js file in line 7.

File: build/media_source/system/js/core.es6/customevent.es6 (hunk 3 above)
Comment (C-Lodder) [reply build/media_source/system/js/core.es6/customevent.es6:2]: Code snippets from MDN after Aug 2010 are licensed under CC0 (https://creativecommons.org/publicdomain/zero/1.0/)

Comment (HLeithner) [reply build/media_source/system/js/core.es6/customevent.es6:2]: Do I have to attribute the person who applied CC0 to their work?

No, there is no legal requirement that you attribute the affirmer, only an expectation that you will voluntarily do so if requested. The CC0 deed provides HTML code that can be copy and pasted into your webpage to easily cite the author and the work, if that information has been provided by the affirmer.


So this is resolved, thx

**Comment by brianteeman** [comment]
@HLeithner can you resolve the conflicts so we can get this merged and resolve the copyright issue please

**Comment by HLeithner** [comment]
I will try to resolve this when I have time... it's a bit more work then expected...

**Comment by wilsonge** [comment]
@HLeithner friendly reminder on this one

**Comment by brianteeman** [comment]
@HLeithner could you resolve the conflicts please so that we can get this merged and resolve the copyright issue asap

**Comment by brianteeman** [comment]
@HLeithner @wilsonge I am marking this as a beta blocker because of the copyright issue

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on dac63318a6c42b1d326a13273c753817499854dd<br /><br />Functionality works as expected after applying the PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26300">issues.joomla.org/tracker/joomla-cms/26300</a>.</sub>

**Comment by jwaisner** [comment]
@HLeithner Drone is failing in MySQL and JS. Can you take a look to see if it is related to your PR?

File: build/media_source/system/js/core.es6.js [hunk 4]
```diff
@@ -115,607 +86,6 @@ window.Joomla.Modal = window.Joomla.Modal || {
 ((Joomla, document) => {
```
Comment (C-Lodder) [inline build/media_source/system/js/core.es6.js]:
```suggestion
((Joomla) => {
```

File: build/media_source/system/js/core.es6.js [hunk 5]
```diff
@@ -742,419 +112,4 @@ window.Joomla.Modal = window.Joomla.Modal || {
     return destination;
   };
```
Comment (C-Lodder) [inline build/media_source/system/js/core.es6.js]:
```suggestion
```

**Review by C-Lodder — CHANGES_REQUESTED** [review]

File: build/media_source/system/js/core.es6.js (hunk 4 above)
Comment (HLeithner) [reply build/media_source/system/js/core.es6.js]: In the last line we call `})(Joomla, document);` document should be removed there too?

Comment (C-Lodder) [reply build/media_source/system/js/core.es6.js]: yup

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on bbe016c34977e22981defcd3b5f8fee9e91308cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26300">issues.joomla.org/tracker/joomla-cms/26300</a>.</sub>

File: build/media_source/system/js/core.es6/ajax.es6 [hunk 6]
```diff
@@ -0,0 +1,123 @@
+/**
+ * @copyright  Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
```
Comment (infograf768) [inline build/media_source/system/js/core.es6/ajax.es6:2]: I guess 2020 in all files concerned

Comment (HLeithner) [reply build/media_source/system/js/core.es6/ajax.es6:2]: This will be updated automatically while building the release

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on bbe016c34977e22981defcd3b5f8fee9e91308cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26300">issues.joomla.org/tracker/joomla-cms/26300</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26300">issues.joomla.org/tracker/joomla-cms/26300</a>.</sub>

**Comment by wilsonge** [comment]
Thankyou for dealing with this @HLeithner !

**Comment by HLeithner** [comment]
Finally YaY thanks!

**Pull request merged**